src/snowflake/snowpark/_internal/data_source_utils.py

```diff
@@ -0,0 +1,347 @@
+        fields.append(
+            StructField(
+                column[0], data_type, True if column[4].lower() == "y" else False
+            )
+        )
+
+    return StructType(fields)
+
+
+def infer_data_source_schema(conn: Connection, table: str) -> StructType:
+    try:
+        current_db = detect_dbms(conn)
+        cursor = conn.cursor()
+        if current_db == DBMS_TYPE.SQL_SERVER_DB:
+            query = f"""
+                    SELECT COLUMN_NAME, DATA_TYPE, NUMERIC_PRECISION, NUMERIC_SCALE, IS_NULLABLE
+                    FROM INFORMATION_SCHEMA.COLUMNS
+                    WHERE TABLE_NAME = '{table}'
+                    """
+            cursor.execute(query)
+            raw_schema = cursor.fetchall()
+            return sql_server_to_snowpark_type(raw_schema)
+        elif current_db == DBMS_TYPE.ORACLE_DB:
+            query = f"""
+                    SELECT COLUMN_NAME, DATA_TYPE, DATA_PRECISION, DATA_SCALE, NULLABLE
+                    FROM USER_TAB_COLUMNS
+                    WHERE TABLE_NAME = '{table}'
+                    """
+            cursor.execute(query)
+            raw_schema = cursor.fetchall()
+            return oracledb_to_snowpark_type(raw_schema)
+        else:
+            raise NotImplementedError(
+                f"currently supported drivers are pyodbc and oracledb, got: {current_db}"
+            )
+    except Exception as exc:
+        raise SnowparkDataframeReaderException(
+            f"Unable to infer schema from table {table}"
+        ) from exc
+
+
+def data_source_data_to_pandas_df(
+    data: List[Any], schema: StructType, conn: Connection
+) -> pd.DataFrame:
+    columns = [col.name for col in schema.fields]
+    df = pd.DataFrame.from_records(data, columns=columns)
+
+    # convert timestamp and date to string to work around SNOW-1911989
+    df = df.map(
+        lambda x: x.isoformat()
+        if isinstance(x, (datetime.datetime, datetime.date))
+        else x
+    )
+    # convert binary type to object type to work around SNOW-1912094
+    df = df.map(lambda x: x.hex() if isinstance(x, (bytearray, bytes)) else x)
+    current_db = detect_dbms(conn)
+    if current_db == DBMS_TYPE.SQL_SERVER_DB or current_db == DBMS_TYPE.SQLITE_DB:
+        return df
+    elif current_db == DBMS_TYPE.ORACLE_DB:
+        from oracledb import LOB
+
+        tz_data = []
+        for col in schema.fields:
+            if isinstance(col.datatype, TimestampType) and col.datatype.tz in [
+                TimestampTimeZone.TZ,
+                TimestampTimeZone.LTZ,
+            ]:
+                tz_data.append(col.name)
+        # apply read to LOB object, we currently have FakeOracleLOB because CLOB and BLOB is represented by an
+        # oracledb object and we cannot add it as our dependency in test, so we fake it in this way
+        # TODO: SNOW-1923698 remove FakeOracleLOB after we have test environment
+        df = df.map(lambda x: x.read() if isinstance(x, (LOB, FakeOracleLOB)) else x)
```

Comment:
I think we should avoid having test related code from the production code.
what you can do is putting the dependencies into a develop requirements, and update tox.ini to have a new env datasource setup, you can check how modin does this (search `modin-development` in the code base).

we can do it in a separate PR.